Blog Post 0005 - Building a Cloud Native Buildpack Isn’t as Hard as You Think (#8)
paketo-buildpacks/blog: content/posts/0004-writing-buildpacks.md

---
title: Building a Cloud Native Buildpack Isn’t as Hard as You Think
date: "2021-02-16"
slug: building-a-cloud-native-buildpack-isnt-as-hard-as-you-think
author: swigmore
---

> Maybe you’ve heard that many companies today are transitioning to, or have already transitioned to, a container-based platform like Kubernetes. With this movement comes the need to containerize your applications, new and old. Buildpacks provide a seamless mechanism for doing that, *without Dockerfiles.* [Paketo Buildpacks](https://paketo.io) is a community-driven project that provides Cloud Native Buildpacks implementations for the most popular languages. Buildpacks provide great out-of-box support for many languages, but what if you need a buildpack that doesn’t exist yet? As a community-driven initiative, it may feel like the onus is on you to create that buildpack. Where would you even begin? 🤯

Reviewer: ```suggestion
Maybe you’ve heard that many companies today are transitioning to, or have already transitioned to, a container-based platform like Kubernetes. With this movement comes the need to containerize your applications, new and old. Buildpacks provide a seamless mechanism for doing that *without Dockerfiles.* [Paketo Buildpacks](https://paketo.io) is a community-driven project that provides [Cloud Native Buildpacks](https://buildpacks.io) implementations for the most popular languages. Buildpacks provide great out-of-box support for many languages, but what if you need a buildpack that doesn’t exist yet? As a community-driven initiative, it may feel like the onus is on you to create that buildpack. Where would you even begin? 🤯
```

This is both a blessing and a curse when it comes to writing buildpacks. It’s a blessing for obvious reasons: since the buildpack only needs to focus on one major task, it makes them simpler to write. We try to pick good defaults that would make the experience better in the vast majority of cases, and to keep our APIs streamlined, but meaningful.

Simple sounds good, how could this be a curse? 
Because our buildpacks perform one task each, in unique use cases it can be hard to discern if you need a fully new buildpack for your case, or if an existing buildpack just needs a tweak. It doesn’t always make sense to architect a new buildpack, but it can be hard to make this judgement if you’re not closely acquainted with buildpacks that are already available.

#### Two roads diverged

One way to determine if you need a new buildpack is to think about if the feature in question is completely new functionality, or if it’s an unimplemented subset of existing buildpack functionality.
![Roads-Diverged](/images/posts/0004/roads-diverged.jpg)

##### When to add a new buildpack

> New functionality usually means a new buildpack is needed. An example of this can be seen with the [Rails Assets buildpack](https://github.com/paketo-buildpacks/rails-assets). A user, [@drnic](https://github.com/drnic) filed [an issue](https://github.com/paketo-buildpacks/ruby/issues/470) about performing asset compilation for Ruby apps, which was completely unsupported behaviour in the Ruby buildpacks. A Ruby maintainer, [@ryanmoran](https://github.com/ryanmoran) wrote up [an RFC](https://github.com/paketo-buildpacks/ruby/pull/475) to propose a new buildpack for the functionality, and a contributor [@genevievelesperance](https://github.com/genevieve) actually [implemented the buildpack](https://github.com/genevieve/rails-assets). (If you’re interested in this workflow, look out for a future blog post outlining our contribution process!) The idea here is that a new buildpack was needed in order to support completely new functionality.

Author: I like this example, I'll update the post to include it.

##### When to modify a buildpack

> On the other hand, there are many cases for modifying an existing buildpack. When looking for cases of “unimplemented subsets of existing functionality”, ask yourself: *If you were building this app again, is there any “optional” behaviour?*

Author: good point

#### Patterns are everywhere

Across the different language ecosystems, the buildpacks follow major patterns depending on the type of job that they are doing. Understanding the patterns that our buildpacks follow can make the decision whether to modify a buildpack or create a new one clearer. Especially in the case of writing a new buildpack, I recommend taking advantage of the work that’s already been done. It's far more helpful to look at what other buildpacks are doing as concrete examples than to try to architect complicated behaviours from scratch. Decide what you need your buildpack to do, and find patterns that suit your case. Here I’ve highlighted some common patterns across our buildpacks:

##### Installing a distribution/tool

> When you need to download some tool or distribution for your app, and (usually) put it on the PATH, you may find yourself saying:

Reviewer: ```suggestion
When you need to download some tool or distribution for your app, and (usually) put it on the PATH, you may find yourself saying:

```

Reviewer: This will fix the formatting so that this section looks like the other "pattern" sections.

Author: Struggled with formatting, thank you!!

* Installation Process Execution: How the command is executed (using `pexec`)
* Dependency Caching: How dependencies and layers are cached/reused for subsequent builds
* Symlinking: Some buildpacks do symlinking between assets and layers, see [Yarn Install](https://github.com/paketo-buildpacks/yarn-install) for an example.

##### Running a start command

When a command needs to be run to start the app in the container, you might think:

*“I'll need to run `go-build` to compile and run my app”* or
*“I have an app that runs on a web server, so I need to run a command to start the server.”*

Check out the following buildpacks: [Go Build](https://github.com/paketo-buildpacks/go-build), [Npm Start](https://github.com/paketo-buildpacks/npm-start), [Dotnet Execute](https://github.com/paketo-buildpacks/dotnet-execute), Ruby webservers (e.g. [Unicorn](https://github.com/paketo-buildpacks/unicorn) or [Passenger](https://github.com/paketo-buildpacks/passenger)). And look for the following things:

> * Detect: What they require during detection to run. This varies widely depending on language.

Author: Hmm. For this one I feel that it does actually vary enough that I wouldn't want to over simplify. I feel like providing different examples to look at, and telling the reader to pay attention to detection criteria which can vary is sufficient here.

#### Bootstrap it

Another huge help is our buildpack [Bootstrapper](https://github.com/paketo-community/bootstrapper) which sets up a basic buildpack directory structure in a `packit`-compliant way. Using the bootstrapper tool removes the guesswork involved in setting up all the basic pieces of a buildpack, and enforces test-driven development. The bootstrapper-provided integration tests fail by default, encouraging you to write tests first and think about the overarching picture of what you need your buildpack to do.

#### Connecting with our community

> My final piece of advice to make contributing a breeze is to connect with our community in any of our forums that suit you. We have a [discussion forum](http://github.com/paketo-buildpacks/feedback/discussions) on Github which is a great place to suggest ideas and get feedback. Additionally, you can file an issue on relevant buildpack repositories, [write an RFC](https://github.com/paketo-buildpacks/rfcs) proposing your idea formally. Lastly, if you fancy a chat we’re available on [Paketo Slack](https://slack.paketo.io) and hold weekly [Paketo Working Group meetings](https://github.com/paketo-buildpacks/community#working-group-meetings)!

Author: I actually made a flow chart for this exact case in my first iteration of this post, but I decided it was TMI for this post. Maybe I can send it to you and get your opinion?

> My final piece of advice to make contributing a breeze is to connect with our community in any of our forums that suit you. We have a [discussion forum](http://github.com/paketo-buildpacks/feedback/discussions) on Github which is a great place to suggest ideas and get feedback. Additionally, you can file an issue on relevant buildpack repositories, [write an RFC](https://github.com/paketo-buildpacks/rfcs) proposing your idea formally. Lastly, if you fancy a chat we’re available on [Paketo Slack](https://slack.paketo.io) and hold weekly [Paketo Working Group meetings](https://github.com/paketo-buildpacks/community#working-group-meetings)!

Reviewer 2: Maybe a flowchart _is_ overkill, but a bullet list of possible next steps at the end would probably suffice. Just something for people to latch onto to get more involved in the community :)

> My final piece of advice to make contributing a breeze is to connect with our community in any of our forums that suit you. We have a [discussion forum](http://github.com/paketo-buildpacks/feedback/discussions) on Github which is a great place to suggest ideas and get feedback. Additionally, you can file an issue on relevant buildpack repositories, [write an RFC](https://github.com/paketo-buildpacks/rfcs) proposing your idea formally. Lastly, if you fancy a chat we’re available on [Paketo Slack](https://slack.paketo.io) and hold weekly [Paketo Working Group meetings](https://github.com/paketo-buildpacks/community#working-group-meetings)!

Reviewer 2: One thing that you could do is split the last paragraph into a list:

Into 
```suggestion
My final piece of advice to make contributing a breeze is to connect with our community in any of our forums that suit you:
    * [Paketo Slack](https://slack.paketo.io): chat with Paketo maintainers in real time to get quick questions answered
    * [Github Discussion forum](http://github.com/paketo-buildpacks/feedback/discussions): suggest ideas and get feedback asynchronously from the Paketo community
    * [RFC process](https://github.com/paketo-buildpacks/rfcs): formally propose your modification
    * [Paketo Working Group meetings](https://github.com/paketo-buildpacks/community#working-group-meetings): bring your ideas, RFCs, buildpacks, etc. for discussion in the weekly video call with Paketo maintainers and contributors
```

> My final piece of advice to make contributing a breeze is to connect with our community in any of our forums that suit you. We have a [discussion forum](http://github.com/paketo-buildpacks/feedback/discussions) on Github which is a great place to suggest ideas and get feedback. Additionally, you can file an issue on relevant buildpack repositories, [write an RFC](https://github.com/paketo-buildpacks/rfcs) proposing your idea formally. Lastly, if you fancy a chat we’re available on [Paketo Slack](https://slack.paketo.io) and hold weekly [Paketo Working Group meetings](https://github.com/paketo-buildpacks/community#working-group-meetings)!

Author: @fg-j ooo I really like this!!! Thank you for laying it out like that, it definitely makes it clearer to read

---
title: Building a Cloud Native Buildpack Isn’t as Hard as You Think
date: "2021-02-05"
slug: building-a-cloud-native-buildpack-isnt-as-hard-as-you-think
author: swigmore
---

Maybe you’ve heard some chatter about Cloud Native Buildpacks, or at least the value in containerizing your applications. Buildpacks provide a seamless mechanism for doing just that. There are a few implementations of the buildpack concept available, but the general takeaways can be summarized as follows:

> “Buildpacks provide a seamless experience from source code to a running container” 

> “Dockerfiles aren’t the only option anymore” 

> “Everything is open source and community driven”

> 

Reviewer 3: I think that this is all valuable information but having it up top lessens the initial impact of the blog post and obfuscates what you are planning on talking about. Which is ultimately the composable nature of buildpacks and how easy it is to write your own.

Author: Regarding this comment and the one below I kind of reworked the intro. See https://github.com/paketo-buildpacks/blog/pull/8/commits/f5ab2645bad7a8962d4b59bbded22925c7dd2934

> This is all very well and good for making containerized applications in a pinch, but what happens when I need a specific buildpack for my use case that doesn’t exist yet? As a community-driven project, is the onus on me to create that buildpack myself? Where does one even begin?

Author: I like this idea

> The Paketo project provides CNB implementations for the most popular languages. Check out https://paketo.io/ to learn more about what's supported. Our buildpacks provide great out-of-box support for most languages, but what if I need a buildpack that doesn’t exist yet? As a community-driven project, is the onus on me to create that buildpack? Where do I even begin?

Reviewer 3: This paragraph is a great intro into what the blog is going to be about. I think that you should make it more prominent in some way.

## Creating a Paketo-style buildpack

There are definitely some need-to-know basics when it comes to buildpacks that have been well documented by the Cloud Native Buildpacks project in [the docs](https://buildpacks.io/docs/concepts/). Major concepts such as the buildpack lifecycle phases, and the other components that work alongside buildpacks are explained there. Once you have the basics down, there is a [simple tutorial](https://buildpacks.io/docs/buildpack-author-guide/create-buildpack/) to create a buildpack in Bash.

### Packit is here to help!

> Kicking things up a notch, the [Creating a Paketo Buildpack tutorial](https://paketo.io/docs/tutorials/create-paketo-buildpack/) on the [Paketo website](paketo.io) outlines how to create a really simple buildpack in **Golang**. With Golang, it’s possible to implement more complex logic with less work than with Bash. A shining star of the tutorial is the introduction of the [Packit library](github.com/paketo-buildpacks/packit), which simplifies the repetitive parts of writing a buildpack, such as [dependency installation](https://github.com/paketo-buildpacks/packit/blob/main/postal) and [file system manipulation](https://github.com/paketo-buildpacks/packit/blob/main/fs). As a part of the Cloud Native Buildpacks project, our buildpacks follow the [CNB Spec](https://github.com/buildpacks/spec). A huge benefit of Packit is that it helps abide by the CNB spec without too much effort.

Reviewer 4: Is Paketo technically `part of the Cloud Native Buildpacks project`?

Reviewer 3: No we are not technically part of the project. We are a standalone project in the CFF utilizing the CNB spec.

Author: Oh ok, whoops. Will fix that!

> Kicking things up a notch, the [Creating a Paketo Buildpack tutorial](https://paketo.io/docs/tutorials/create-paketo-buildpack/) on the [Paketo website](paketo.io) outlines how to create a really simple buildpack in **Golang**. With Golang, it’s possible to implement more complex logic with less work than with Bash. A shining star of the tutorial is the introduction of the [Packit library](github.com/paketo-buildpacks/packit), which simplifies the repetitive parts of writing a buildpack, such as [dependency installation](https://github.com/paketo-buildpacks/packit/blob/main/postal) and [file system manipulation](https://github.com/paketo-buildpacks/packit/blob/main/fs). As a part of the Cloud Native Buildpacks project, our buildpacks follow the [CNB Spec](https://github.com/buildpacks/spec). A huge benefit of Packit is that it helps abide by the CNB spec without too much effort.

Reviewer 5: Maybe move the first couple sentences here into the intro paragraph before so it flows better into Packit section. So something like "...create a buildpack in Bash. A buildpack written in Bash works great for a simple buildpack, but the Paketo contributors have adopted Golang as our language of choice to..."

> Writing a buildpack with the Packit library is integral to how we write our buildpacks and keep them consistent. I recommend checking out Packit, and then also looking at one of our buildpacks to see how the packages are used. Key an eye out for future blog post deep diving into Packit as well.

Reviewer 4: s/Key/Keep

### Bootstrap it

> Another huge help is our buildpack [Bootstrapper](https://github.com/paketo-community/bootstrapper) which sets up a basic buildpack directory structure in a Packit-compliant way. Using the bootstrapper tool removes the guesswork involved in setting up all the basic pieces of a buildpack, and enforces test-driven development. The bootstrapper-provided integration tests fail by default, encouraging you to write tests first and think about the overarching picture of what you need your buildpack to do.

Reviewer 5: ```suggestion
Another huge help is our buildpack [Bootstrapper](https://github.com/paketo-community/bootstrapper) which sets up a basic buildpack directory structure in a Packit-compliant way. Using the bootstrapper tool removes the guesswork involved in setting up all the basic pieces of a buildpack, and lets you test the buildpack along the way. The bootstrapper-provided integration tests fail by default, encouraging you to write tests first and think about the overarching picture of what you need your buildpack to do.
```
"Enforces" feels a bit strong here

Author: I want to mention test-driven development still. How about:
"and allows you the option of test driven development"

To containerize a simple Go application, we need a couple things: a relevant Golang version, some way to manage packages (such as Go modules, maybe), and a way to compile and run the app. Depending on the specific needs of the app, each of these steps is handled by one implementation buildpack, which are pulled in by the language family buildpack. 

![Go-Implementation](/images/posts/0004/go-implementation.jpg)

### **Patterns are everywhere**

One of the best pieces of advice I can provide is to take advantage of the work that’s already been done. It's far more helpful to look at what other buildpacks are doing as concrete examples than to try to architect complicated behaviours from scratch or based on a basic tutorial. Across the different language ecosystems, the buildpacks follow major patterns depending on the type of job that they are doing. Decide what you need your buildpack to do, and find patterns that suit your case. Here I’ve highlighted some common patterns across our buildpacks:

> ### Pattern 1: Installing a distribution/tool

Reviewer 5: ```suggestion
### Installing a distribution/tool
```

When you need to download some tool or distribution for your app, and (usually) put it on the PATH. Think:
> “I'll need Golang to run my Go app.”,\
> “I'll need to install Bundler to run `bundle-install` commands for my Ruby app."

Check out the following buildpacks: [Go Dist](https://github.com/paketo-buildpacks/go-dist) , [Bundler](https://github.com/paketo-buildpacks/bundler), [Yarn](https://github.com/paketo-buildpacks/yarn), [Node Engine](https://github.com/paketo-buildpacks/node-engine)

Key things to look out for:
* Detection: Variations of checking language-specific files (e.g. Gemfile.lock) or environment variables for versions to install 
* Build: How layer metadata is set<br/></br> 

> ### Pattern 2: Managing/installing app dependencies

Reviewer 5: ```suggestion
### Managing/installing app dependencies
```

When your app has a dependency manager, this pattern is needed to actually run that command to pull in dependencies. Think: 
> “I'll need to run `bundle-install` to pull in certain gems for my Ruby app.”\
> “I'll need to run `go mod vendor` to vendor my Go app dependencies.”

Check out the following buildpacks: [Bundle Install](https://github.com/paketo-buildpacks/bundle-install), [Go Mod Vendor](https://github.com/paketo-buildpacks/go-mod-vendor), [Yarn Install](https://github.com/paketo-buildpacks/yarn-install), [Rails Assets](github.com/paketo-buildpacks/rails-assets)

Key things to look out for:
* Installation Process Execution: How the command is executed (using pexec)
* Dependency Caching: How dependencies and layers are cached/checked for subsequent builds
* Symlinking: Some buildpacks do symlinking between assets and layers, see [Yarn Install](https://github.com/paketo-buildpacks/yarn-install) for an example.

> ### Pattern 3: Running a start command

Reviewer 5: ```suggestion
### Running a start command
```
